### PR TITLE
import pois type node

### PR DESCRIPTION
**Pull request by datanel**
### TODO

- [x] add poi type
- [x] add tests
- [x] add doc

### CODE REVIEW

File: src/bin/osm2mimir.rs [hunk 1]
```diff
@@ -251,6 +253,18 @@ fn get_street_admin(admins_geofinder: &AdminGeoFinder,
     coord.map_or(vec![], |c| admins_geofinder.get(&c))
 }
 
+fn get_poi_admin(admins_geofinder: &AdminGeoFinder,
+                    node: &osmpbfreader::objects::Node)
+                    -> Vec<Rc<mimir::Admin>> {
+    let coord = Some(geo::Coordinate {
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: why wrap this in an Option ?

File: src/bin/osm2mimir.rs [hunk 2]
```diff
@@ -389,6 +403,62 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec, city_level: u32) -> Stree
     street_list
 }
 
+type TypesVec = BTreeMap<String, BTreeMap<String, String>>;
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: maybe a more explicit type would be better

File: src/bin/osm2mimir.rs [hunk 3]
```diff
@@ -389,6 +403,62 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec, city_level: u32) -> Stree
...
+impl PoiMatcher {
+    pub fn new(types: TypesVec) -> PoiMatcher {
+        PoiMatcher { poi_types: types }
+    }
+
+    pub fn is_poi(&self, obj: &osmpbfreader::OsmObj) -> bool {
+        match *obj {
+            osmpbfreader::OsmObj::Node(ref node) => {
+                let mut ret = false;
+                for (poi_tag, poi_types) in &self.poi_types {
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: you can rewrite this with iterator and `any` :stuck_out_tongue_winking_eye:

File: src/bin/osm2mimir.rs [hunk 4]
```diff
@@ -389,6 +403,62 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec, city_level: u32) -> Stree
...
+            let admins = get_poi_admin(&admins_geofinder, node);
+
+            let poi = mimir::Poi {
+                id: format!("poi:osm:{}", node.id).to_string(),
+                name: node_name.to_string(),
+                label: format_label(&admins, city_level, node_name),
+                coord: mimir::Coord::new(node.lat, node.lon),
+                administrative_regions: admins,
+            };
+            pois.push(poi);
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: you can rewrite this with iterators, `filter` and `collect` :wink:

File: src/bin/osm2mimir.rs [hunk 5]
```diff
@@ -434,4 +504,35 @@ fn main() {
                           .unwrap();
     info!("Nb of indexed admin: {}", nb_admins);
 
+    if args.flag_import_poi {
+        let mut poi_types = TypesVec::new();
+        let mut amenity_types = BTreeMap::new();
+        amenity_types.insert("university".to_string(), "université".to_string());
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: I think this should go in a separate function for clarity purpose

File: src/bin/osm2mimir.rs (hunk 2 above)
Comment (datanel) [inline src/bin/osm2mimir.rs]: like what?

File: src/bin/osm2mimir.rs (hunk 3 above)
Comment (datanel) [inline src/bin/osm2mimir.rs]: mmmh, well I make something that works I will see after for iterators

File: src/bin/osm2mimir.rs [hunk 6]
```diff
@@ -389,6 +403,62 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec, city_level: u32) -> Stree
...
+                            ret = true;
+                            break;
+                        }
+                    }
+                }
+                ret
+
+            }
+            _ => false,
+        }
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Not tested, but you have the idea.

```rust
obj.is_node() && self.poi_types.iter().filter_map(|(poi_tag, poi_types)| {
    obj.tags().get(poi_tag).and_then(|poi_type| poi_types.contains_key(poi_type))
}).any()
```

File: src/bin/osm2mimir.rs (hunk 5 above)
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
fn default_amenity_types() -> BTreeMap<String, String> {
    [
        ("university", "université"),
        ("hospital", "hôpital"),
        ("post_office", "bureau de poste"),
        ("bicycle_rental", "station vls"),
        ("bicycle_parking", "Parking vélo"),
        ("parking", "Parking"),
        ("police", "Police, Gendarmerie"),
        ("townhall", "Mairie")
    ].iter().map(|&(k, v)| (k.to_string(), v.to_string())).collect()
}
```

File: src/bin/osm2mimir.rs [hunk 7]
```diff
@@ -434,4 +504,35 @@ fn main() {
...
+        poi_types.insert("amenity".to_string(), amenity_types);
+        poi_types.insert("leisure".to_string(), leisure_types);
+
+        info!("Extracting pois from osm");
+        let pois = pois(&mut parsed_pbf, poi_types, &admins, city_level);
+
+        info!("Importing pois into Mimir");
+        for poi in &pois {
+            println!("{:?}", poi);
+        }
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: debug to remove

Comment (datanel) [inline src/bin/osm2mimir.rs]: i know

File: src/bin/osm2mimir.rs (hunk 5 above)
Comment (datanel) [inline src/bin/osm2mimir.rs]: and the same for leisure?

File: src/bin/osm2mimir.rs (hunk 2 above)
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: At least PoiTypes is better, first because it's more clear that's poi related, and also because that's not a vec ;-)

Comment (datanel) [inline src/bin/osm2mimir.rs]: lol it was a vec before but i didn't change the name. thanks

Comment (datanel) [inline src/bin/osm2mimir.rs]: done

File: src/bin/osm2mimir.rs (hunk 5 above)
Comment (datanel) [inline src/bin/osm2mimir.rs]: done

File: src/bin/osm2mimir.rs (hunk 7 above)
Comment (datanel) [inline src/bin/osm2mimir.rs]: done

File: src/bin/osm2mimir.rs [hunk 8]
```diff
@@ -389,6 +403,107 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec, city_level: u32) -> Stree
...
 
+type PoiTypes = BTreeMap<String, BTreeMap<String, String>>;
+
+#[derive(Debug)]
+struct PoiMatcher {
+    poi_types: PoiTypes,
+}
+
+#[derive(Debug)]
+struct Tag {
```
Comment (datanel) [inline src/bin/osm2mimir.rs]: need a better name

File: src/bin/osm2mimir.rs [hunk 9]
```diff
@@ -389,6 +403,107 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec, city_level: u32) -> Stree
...
+        obj.is_node() && self.poi_types.iter().any(|(poi_tag, poi_types)| {
+            obj.tags().get(poi_tag).map_or(false, |poi_type| {
+                poi_types.contains_key(poi_type)
+            })
+        })
+    }
+
+    pub fn get_poi_type(&self, node: &osmpbfreader::Node) -> Option<Tag> {
+
+        for (poi_tag, poi_types) in &self.poi_types {
```
Comment (datanel) [inline src/bin/osm2mimir.rs]: will be done with iterators but need some help

File: libs/mimir/src/objects.in.rs [hunk 10]
```diff
@@ -122,6 +122,34 @@ impl<'a, T: EsId> EsId for Rc<T> {
...
+}
+
+impl EsId for Poi {
+    fn es_id(&self) -> Option<String> {
+        Some(self.id.clone())
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct PoiType {
```
Comment (pbougue) [inline libs/mimir/src/objects.in.rs]: Not sure we should use the same structure as in jormun

File: src/bin/osm2mimir.rs [hunk 11]
```diff
@@ -389,6 +403,93 @@ fn streets(pbf: &mut OsmPbfReader, admins: &AdminsVec, city_level: u32) -> Stree
...
+        }).map(|(poi_tag, poi_types)| {
+            let tag_value = node.tags.get(poi_tag).unwrap();
+            let tag_value_translation = poi_types.get(tag_value).unwrap();
+
+            Tag {
+                tag_name: poi_tag.clone(),
+                tag_value: tag_value.clone(),
+                tag_value_translation: tag_value_translation.clone()
+            }
+        }).unwrap()
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: I don't like this unwrap. Nothing can say that the given node will have a Tag.

Comment (datanel) [inline src/bin/osm2mimir.rs]: this method is called after `let objects = osmpbfreader::get_objs_and_deps(pbf, |o| matcher.is_poi(o)).unwrap();` which filters pois with tags amenity, ... so node here has a tag

Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Yeah, but nothing in the impl implies that. I'd prefer an unwrap at the upper level

Comment (datanel) [inline src/bin/osm2mimir.rs]: https://github.com/CanalTP/mimirsbrunn/pull/86/files#diff-8be6b8fd629e2b03d7a46cd92d72be2bR458 here?

Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: yep

File: libs/mimir/src/objects.in.rs [hunk 12]
```diff
@@ -52,19 +52,19 @@ pub enum Place {
 }
 
 impl Place{
-    pub fn is_admin(&self) -> bool { 
+    pub fn is_admin(&self) -> bool {
```
Comment (datanel) [inline libs/mimir/src/objects.in.rs]: done

File: json/settings.json [hunk 13]
```diff
@@ -179,6 +179,46 @@
...
+                        "id": { "type": "string", "index": "no" },
+                        "level": { "type": "long", "index": "no" },
+                        "zip_codes": { "type": "string", "index": "not_analyzed" },
+                        "weight": { "type": "long" }
+                    }
+                },
+                "poi_type": {
+                    "properties": {
+                        "id": { "type": "string", "index": "no" },
+                        "name": { "type": "string" }
```
Comment (pbougue) [inline json/settings.json]: That question remains... @TeXitoi any opinion?

File: tests/osm2mimir_bano2mimir_test.rs [hunk 14]
```diff
@@ -49,17 +50,17 @@ pub fn osm2mimir_bano2mimir_test(es_wrapper: ::ElasticSearchWrapper) {
                              format!("--connection-string={}", es_wrapper.host())],
                         &es_wrapper);
 
-    // after an import, we should have 3 index, and some aliases to this index
+    // after an import, we should have 4 index, and some aliases to this index
```
Comment (pbougue) [inline tests/osm2mimir_bano2mimir_test.rs]: index`es`

File: tests/osm2mimir_test.rs [hunk 15]
```diff
@@ -78,4 +79,29 @@ pub fn osm2mimir_sample_test(es_wrapper: ::ElasticSearchWrapper) {
...
+    // Test: search Pois by label
+    let res: Vec<_> = es_wrapper.search_and_filter("label:Le-Mée-sur-Seine Courtilleraies", |_| true).collect();
+    assert!(res.len() != 0);
+    assert!(res.iter().any(|r| r.is_poi()));
+
+    let res: Vec<_> = es_wrapper.search_and_filter("label:Melun Rp", |_| true).collect();
+    assert!(res.len() != 0);
+    assert!(res.iter().any(|r| r.is_poi()));
+
+    let res: Vec<_> = es_wrapper.search_and_filter("bureau de poste", |_| true).collect();
```
Comment (pbougue) [inline tests/osm2mimir_test.rs]: Not sure this is usefull, as we have the next test?

File: tests/osm2mimir_test.rs [hunk 16]
```diff
@@ -78,4 +79,29 @@ pub fn osm2mimir_sample_test(es_wrapper: ::ElasticSearchWrapper) {
...
+    // Test: search Pois of type "bureau de poste"
+    let res: Vec<_> = es_wrapper.search_and_filter("bureau de poste", |_| true)
+        .filter(|place: &mimir::Place| {
+            place.is_poi()
+        })
+        .collect();
+    assert_eq!(res.len(), 2);
+
+    // Test: search Poi of type "université"
+    let nb = es_wrapper.search_and_filter("université", |_| true).count();
```
Comment (pbougue) [inline tests/osm2mimir_test.rs]: This is to test that we can't read way-POI? If it's the case it's not really useful as it's not a targeted behaviour (but at least we will have a test -to change- when we code that feature)

File: tests/osm2mimir_test.rs (hunk 15 above)
Comment (datanel) [inline tests/osm2mimir_test.rs]: good point

File: json/settings.json [hunk 17]
```diff
@@ -179,6 +179,47 @@
...
+                        "level": { "type": "long", "index": "no" },
+                        "zip_codes": { "type": "string", "index": "not_analyzed" },
+                        "weight": { "type": "long" }
+                    }
+                },
+                "weight": { "type": "long" },
+                "poi_type": {
+                    "properties": {
+                        "id": { "type": "string", "index": "no" },
+                        "name": { "type": "string" }
```
Comment (pbougue) [inline json/settings.json]: seen with @TeXitoi : let's do like the label for that

Comment (pbougue) [inline json/settings.json]: Actually just remove the POI type from ES for now

File: libs/mimir/src/objects.in.rs [hunk 18]
```diff
@@ -122,6 +131,44 @@ impl<'a, T: EsId> EsId for Rc<T> {
...
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Poi {
+    pub id: String,
+    pub label: String,
+    pub name: String,
+    pub coord: Coord,
+    pub administrative_regions: Vec<Rc<Admin>>,
+    pub weight: u32,
+    pub poi_type: PoiType,
```
Comment (pbougue) [inline libs/mimir/src/objects.in.rs]: To be removed as we don't need it in base for now

File: libs/mimir/src/objects.in.rs (hunk 10 above)
Comment (pbougue) [inline libs/mimir/src/objects.in.rs]: To be removed also

File: libs/bragi/src/model.in.rs [hunk 19]
```diff
@@ -187,6 +189,28 @@ impl From<mimir::Addr> for GeocodingResponse {
...
 }
 
+impl From<mimir::Poi> for GeocodingResponse {
+    fn from(other: mimir::Poi) -> GeocodingResponse {
+        let type_ = "poi".to_string();
+        let name = Some(other.label);
+        let label = name.clone();
+        let admins = other.administrative_regions;
+        let city = get_city_name(&admins);
+
```
Comment (pbougue) [inline libs/bragi/src/model.in.rs]: I think you can add postcode the same way it's done in address: `let postcode = if other.zip_codes.is_empty() { None } else { Some(other.zip_codes.join(";")) };`

File: libs/bragi/src/model.in.rs [hunk 20]
```diff
@@ -187,6 +189,28 @@ impl From<mimir::Addr> for GeocodingResponse {
     }
 }
 
+impl From<mimir::Poi> for GeocodingResponse {
+    fn from(other: mimir::Poi) -> GeocodingResponse {
+        let type_ = "poi".to_string();
+        let name = Some(other.label);
```
Comment (pbougue) [inline libs/bragi/src/model.in.rs]: looks like it's inverted, should probably look like address

**Pull request merged**